File: Quicksort/sort.php
<?php 

function lesserAndGreater(array $ary, int $pivot): array
{
  $lesser = [];
  $greater = [];
  foreach ($ary as $value) {
    if ($value < $pivot)
      $lesser[] = $value;
    else
      $greater[] = $value;
  }
  return [$lesser, $greater];
}

function qsort(array $ary): array
{
  if (empty($ary)) return $ary;
  $pivot = array_shift($ary);
  [$lesser, $greater] = lesserAndGreater($ary, $pivot);
  return array_merge(qsort($lesser), [$pivot], qsort($greater));
}

$a = [];
foreach (range(1, 20) as $i)
  $a[] = rand(1, 100);

print_r($a);
print_r(qsort($a));

File: QuickSort/sort.go
package main

import (
	"fmt"
	"math/rand"
)

func main() {
	ary := make([]int, 100)
	for i := range ary {
		ary[i] = rand.Intn(1000)
	}
	fmt.Println(ary)
	fmt.Println(qsort(ary))
}

func qsort(ary []int) []int {
	if len(ary) == 0 {
		return ary
	}
	pivot, ary := ary[0], ary[1:]
	lesser, greater := lesserAndGreater(ary, pivot)
	lesser, greater = qsort(lesser), qsort(greater)
	return append(append(lesser, pivot), greater...)
}

func lesserAndGreater(ary []int, pivot int) (lesser []int, greater []int) {
	for i, v := range ary {
		if ary[i] > pivot {
			greater = append(greater, v)
		} else {
			lesser = append(lesser, v)
		}
	}
	return lesser, greater
}
